File: test_tennis_apis.py
#!/usr/bin/env python3
"""
Test script to evaluate different Tennis APIs for Bragging Rights
"""

import requests
import json
from datetime import datetime

def test_espn_tennis():
    """Test ESPN Tennis API - our primary choice"""
    print("\n1. Testing ESPN Tennis API")
    print("-" * 40)
    
    try:
        print("📍 Base URL: https://site.api.espn.com/apis/site/v2/sports/tennis")
        print("💰 Pricing: FREE (no key required)")
        
        # Test scoreboard endpoint
        response = requests.get(
            "https://site.api.espn.com/apis/site/v2/sports/tennis/scoreboard",
            timeout=5
        )
        
        if response.status_code == 200:
            data = response.json()
            events = data.get('events', [])
            print(f"✅ API Working! Live/Recent matches: {len(events)}")
            
            if events:
                match = events[0]
                competition = match.get('competitions', [{}])[0]
                competitors = competition.get('competitors', [])
                if len(competitors) >= 2:
                    p1 = competitors[0].get('athlete', {}).get('displayName', 'Unknown')
                    p2 = competitors[1].get('athlete', {}).get('displayName', 'Unknown')
                    status = competition.get('status', {}).get('type', {}).get('description', '')
                    print(f"   Sample match: {p1} vs {p2}")
                    print(f"   Status: {status}")
        
        # Test ATP rankings
        print("\n   Testing ATP Rankings...")
        rankings_response = requests.get(
            "https://site.api.espn.com/apis/site/v2/sports/tennis/atp/rankings",
            timeout=5
        )
        
        if rankings_response.status_code == 200:
            rankings_data = rankings_response.json()
            if 'rankings' in rankings_data:
                print("   ATP Rankings available")
                top_3 = rankings_data['rankings'].get('competitors', [])[:3]
                for player in top_3:
                    athlete = player.get('athlete', {})
                    rank = player.get('rank', 'N/A')
                    name = athlete.get('displayName', 'Unknown')
                    print(f"   #{rank}: {name}")
        
        # Test WTA rankings
        print("\n   Testing WTA Rankings...")
        wta_response = requests.get(
            "https://site.api.espn.com/apis/site/v2/sports/tennis/wta/rankings",
            timeout=5
        )
        
        if wta_response.status_code == 200:
            print("   WTA Rankings available")
        
        # Test tournaments
        print("\n   Testing Tournament Data...")
        tournament_response = requests.get(
            "https://site.api.espn.com/apis/site/v2/sports/tennis/scoreboard?limit=50",
            timeout=5
        )
        
        if tournament_response.status_code == 200:
            data = tournament_response.json()
            leagues = data.get('leagues', [])
            if leagues:
                print(f"   Tournaments found: {len(leagues)}")
                for league in leagues[:3]:
                    print(f"   - {league.get('name', 'Unknown')}")
        
        print("\nESPN Tennis API Summary:")
        print("   - Live scores: YES")
        print("   - Rankings: YES")
        print("   - Tournaments: YES")
        print("   - No API key: YES")
        print("   - Free to use: YES")
        
    except Exception as e:
        print(f"ERROR: {e}")

def test_tennis_live_data():
    """Test Tennis Live Data API (RapidAPI)"""
    print("\n2. Testing Tennis Live Data API (RapidAPI)")
    print("-" * 40)
    
    print("   Base URL: https://tennis-live-data.p.rapidapi.com")
    print("   Pricing: Free tier (100 req/month), Pro ($29/month)")
    print("   Requires RapidAPI key")
    print("   Features: Live scores, H2H, detailed stats")
    print("   Limitations: Low free tier limit")

def test_api_tennis():
    """Test API-Tennis"""
    print("\n3. Testing API-Tennis")
    print("-" * 40)
    
    print("   Base URL: https://api-tennis.com")
    print("   Pricing: Free with attribution")
    print("   Features: Live scores, rankings, H2H")
    print("   Requires registration")

def test_flashscore():
    """Test Flashscore Tennis (web scraping)"""
    print("\n4. Testing Flashscore Tennis")
    print("-" * 40)
    
    print("   Base URL: https://www.flashscore.com/tennis/")
    print("   Pricing: Free (web scraping)")
    print("   Features: Comprehensive live scores, odds")
    print("   Limitations: Requires web scraping")

def test_sofascore():
    """Test SofaScore API"""
    print("\n5. Testing SofaScore API")
    print("-" * 40)
    
    try:
        # SofaScore has undocumented API endpoints
        response = requests.get(
            "https://api.sofascore.com/api/v1/sport/tennis/events/live",
            headers={'User-Agent': 'Mozilla/5.0'},
            timeout=5
        )
        
        if response.status_code == 200:
            data = response.json()
            events = data.get('events', [])
            print("   Undocumented API working!")
            print(f"   Live matches: {len(events)}")
        else:
            print("   API returned status:", response.status_code)
            
    except Exception as e:
        print(f"   Connection failed: {e}")
    
    print("   Base URL: https://api.sofascore.com")
    print("   Pricing: Free (undocumented)")
    print("   Features: Live scores, detailed stats, H2H")
    print("   Limitations: Undocumented, may change")

def print_recommendation():
    """Print final recommendation"""
    print("\n" + "=" * 60)
    print("RECOMMENDATION FOR BRAGGING RIGHTS")
    print("=" * 60)
    
    print("""
PRIMARY IMPLEMENTATION: ESPN Tennis API
----------------------------------------
- Completely FREE with no API key
- Same pattern as NBA/NFL/NHL/MLB
- Reliable and stable
- Good coverage of major tournaments
- Live scores and rankings

DATA COVERAGE:
- Live match scores: YES
- Daily schedules: YES
- ATP/WTA rankings: YES
- Tournament info: YES
- Basic match stats: YES
- Player profiles: Partial
- H2H records: NO (need alternative)
- Surface stats: NO (need alternative)

IMPLEMENTATION PLAN:
1. Use ESPN Tennis as primary source
2. Follow same pattern as other ESPN sports
3. Mock H2H and surface stats for MVP
4. Add enhanced data sources later if needed

EDGE INTELLIGENCE FOR TENNIS:
- Current form (last 5 matches)
- Head-to-head history (mock initially)
- Surface preference (hard/clay/grass)
- Tournament importance
- Ranking differential
- Recent injuries/news
    """)

def main():
    print("Testing Tennis APIs for Bragging Rights Integration")
    print("=" * 60)
    
    # Test each API
    test_espn_tennis()
    test_tennis_live_data()
    test_api_tennis()
    test_flashscore()
    test_sofascore()
    
    # Print recommendation
    print_recommendation()
    
    print("\nTesting Complete!")

if __name__ == "__main__":
    main()

File: test_tennis_apis.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Test script to evaluate different Tennis APIs
/// Run with: dart test_tennis_apis.dart
void main() async {
  print('🎾 Testing Tennis APIs for Bragging Rights Integration\n');
  print('=' * 60);
  
  // Test each API
  await testTennisLiveDataAPI();
  await testSportRadarAPI();
  await testAPITennis();
  await testTennisDataAPI();
  await testUltimateTennisAPI();
  
  print('\n' + '=' * 60);
  print('📊 API Evaluation Complete!');
}

/// 1. Tennis Live Data API (RapidAPI)
Future<void> testTennisLiveDataAPI() async {
  print('\n1️⃣ Testing Tennis Live Data API (RapidAPI)');
  print('-' * 40);
  
  try {
    // Free tier: 100 requests/month
    final endpoints = [
      '/matches/live',
      '/matches/today',
      '/players/atp-ranking',
      '/players/wta-ranking',
      '/tournaments/active',
    ];
    
    print('📍 Base URL: https://tennis-live-data.p.rapidapi.com');
    print('💰 Pricing: Free tier (100 req/month), Pro ($29/month)');
    print('📋 Endpoints to test:');
    for (var endpoint in endpoints) {
      print('   • $endpoint');
    }
    
    // Test without key for now
    print('⚠️  Requires RapidAPI key - sign up at rapidapi.com');
    print('✅ Features: Live scores, rankings, H2H, tournaments');
    print('✅ Coverage: ATP, WTA, ITF, Challengers');
    print('❌ Limitations: Low free tier limit');
    
  } catch (e) {
    print('❌ Error: $e');
  }
}

/// 2. SportRadar Tennis API
Future<void> testSportRadarAPI() async {
  print('\n2️⃣ Testing SportRadar Tennis API');
  print('-' * 40);
  
  try {
    print('📍 Base URL: https://api.sportradar.com/tennis/trial/v3/en');
    print('💰 Pricing: Trial (1000 req/month), Enterprise pricing');
    
    // Test trial endpoint (no key needed for basic info)
    final response = await http.get(
      Uri.parse('https://api.sportradar.com/tennis/trial/v3/en/schedules/live/schedule.json?api_key=TRIAL_KEY'),
    ).timeout(Duration(seconds: 5));
    
    if (response.statusCode == 403) {
      print('⚠️  Requires API key from sportradar.com');
    } else if (response.statusCode == 200) {
      final data = json.decode(response.body);
      print('✅ Live matches available: ${data['sport_events']?.length ?? 0}');
    }
    
    print('✅ Features: Comprehensive stats, live data, push feeds');
    print('✅ Coverage: All major tournaments');
    print('❌ Limitations: Expensive for production');
    
  } catch (e) {
    print('⚠️  Connection test failed (expected without key)');
  }
}

/// 3. API-Tennis (Alternative free option)
Future<void> testAPITennis() async {
  print('\n3️⃣ Testing API-Tennis');
  print('-' * 40);
  
  try {
    print('📍 Base URL: https://api-tennis.com/tennis/');
    print('💰 Pricing: FREE with attribution');
    
    // Test public endpoint
    final response = await http.get(
      Uri.parse('https://api-tennis.com/tennis/?method=get_matches&APIkey=FREE_TRIAL'),
    ).timeout(Duration(seconds: 5));
    
    if (response.statusCode == 200) {
      print('✅ API is accessible');
      final body = response.body;
      if (body.contains('error')) {
        print('⚠️  Requires registration at api-tennis.com');
      }
    }
    
    print('✅ Features: Live scores, rankings, H2H');
    print('✅ Coverage: ATP, WTA, ITF');
    print('✅ Free tier available with attribution');
    print('❌ Limitations: Basic data only');
    
  } catch (e) {
    print('⚠️  Connection failed: $e');
  }
}

/// 4. Tennis Data API (tennis-data.co.uk)
Future<void> testTennisDataAPI() async {
  print('\n4️⃣ Testing Tennis Data API');
  print('-' * 40);
  
  try {
    print('📍 Base URL: https://www.tennis-data.co.uk');
    print('💰 Pricing: FREE (CSV data)');
    
    // Test data availability
    final response = await http.get(
      Uri.parse('https://www.tennis-data.co.uk/2024/2024.xlsx'),
    ).timeout(Duration(seconds: 5));
    
    if (response.statusCode == 200) {
      print('✅ Historical data available');
      print('📊 Format: CSV/Excel files');
    } else {
      print('⚠️  Data format: CSV files (not REST API)');
    }
    
    print('✅ Features: Historical match data, odds');
    print('✅ Coverage: All major tournaments');
    print('❌ Limitations: Not real-time, CSV format only');
    
  } catch (e) {
    print('⚠️  Connection test: $e');
  }
}

/// 5. Ultimate Tennis Statistics API
Future<void> testUltimateTennisAPI() async {
  print('\n5️⃣ Testing Ultimate Tennis Statistics');
  print('-' * 40);
  
  try {
    print('📍 Base URL: https://www.ultimatetennisstatistics.com');
    print('💰 Pricing: FREE (web scraping required)');
    
    // Test if site is accessible
    final response = await http.get(
      Uri.parse('https://www.ultimatetennisstatistics.com/tournamentEvents'),
    ).timeout(Duration(seconds: 5));
    
    if (response.statusCode == 200) {
      print('✅ Website accessible');
      print('📊 Data available via web scraping');
    }
    
    print('✅ Features: Detailed statistics, H2H, rankings');
    print('✅ Coverage: Historical and current data');
    print('❌ Limitations: Requires web scraping');
    
  } catch (e) {
    print('⚠️  Connection test: $e');
  }
}

/// Test ESPN Tennis endpoints
Future<void> testESPNTennis() async {
  print('\n6️⃣ Testing ESPN Tennis API');
  print('-' * 40);
  
  try {
    print('📍 Base URL: https://site.api.espn.com/apis/site/v2/sports/tennis');
    print('💰 Pricing: FREE (no key required)');
    
    // Test scoreboard endpoint
    final response = await http.get(
      Uri.parse('https://site.api.espn.com/apis/site/v2/sports/tennis/scoreboard'),
    ).timeout(Duration(seconds: 5));
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      final events = data['events'] ?? [];
      print('✅ API Working! Live matches: ${events.length}');
      
      if (events.isNotEmpty) {
        final match = events[0];
        final competition = match['competitions']?[0];
        if (competition != null) {
          final competitors = competition['competitors'] ?? [];
          if (competitors.length >= 2) {
            print('📍 Sample match: ${competitors[0]['athlete']['displayName']} vs ${competitors[1]['athlete']['displayName']}');
          }
        }
      }
    }
    
    // Test rankings
    final rankingsResponse = await http.get(
      Uri.parse('https://site.api.espn.com/apis/site/v2/sports/tennis/atp/rankings'),
    ).timeout(Duration(seconds: 5));
    
    if (rankingsResponse.statusCode == 200) {
      print('✅ ATP Rankings available');
    }
    
    print('✅ Features: Live scores, rankings, tournaments');
    print('✅ Coverage: All major tournaments');
    print('✅ No API key required!');
    print('❌ Limitations: Limited player stats');
    
  } catch (e) {
    print('❌ Error: $e');
  }
}

/// Summary and Recommendation
void printRecommendation() {
  print('\n' + '=' * 60);
  print('🏆 RECOMMENDATION FOR BRAGGING RIGHTS');
  print('=' * 60);
  
  print('''
  
PRIMARY CHOICE: ESPN Tennis API
--------------------------------
✅ FREE with no API key required
✅ Reliable (same as NBA/NFL/NHL integration)
✅ Live scores and schedules
✅ ATP/WTA rankings
✅ Tournament coverage
✅ Already familiar with ESPN API structure

BACKUP OPTION: API-Tennis
-------------------------
✅ Free tier available
✅ Good coverage
✅ REST API format
❌ Requires registration

IMPLEMENTATION PLAN:
1. Use ESPN Tennis as primary data source
2. Implement same pattern as other ESPN sports
3. Add API-Tennis as fallback if needed
4. Focus on:
   - Live match scores
   - Daily schedules
   - Rankings (ATP/WTA)
   - Tournament information
   - Basic player stats

MISSING DATA (need alternative sources):
- Detailed H2H records
- Surface-specific stats
- Advanced player metrics
- Weather conditions for outdoor matches
  ''');
}